dockertest/dockercmd.py

```diff
@@ -334,6 +334,49 @@ def execute(self, stdin=None):
                                          stdin=stdin, close_fds=True)
         return self.cmdresult
 
+    def wait_for_ready(self, timeout=None, timestep=0.2):
+        """
+        Monitor the output of a container (including docker logs, in
+        case stdout is detached), waiting for the string 'READY'. Return
+        as soon as we see it. If we don't, throw a meaningful exception.
+
+        :raises DockerExecError: on timeout.
+        """
+        end_time = time.time() + timeout
+        while time.time() <= end_time:
+            time.sleep(timestep)
+            if 'READY' in self.stdout:
+                return
+            # Also check docker logs
+            cid = self.container_id
+            if cid is not None:
+                logs = DockerCmd(self.subtest, 'logs', [self.container_id])
+                logs.execute()
+                if 'READY' in logs.stdout:
+                    return
+
+        # Never saw READY. Did container exit? If so, help user understand why
+        if self.done:
+            msg = "Container exited before READY"
+            if self.exit_status == 0:
+                msg += " (normal exit status)"
+            else:
+                msg += "; exit status = %d" % self.exit_status
+            stderr = self.stderr
+            if stderr:
+                msg += "; stderr='%s'" % stderr
+            raise DockerExecError(msg)
+
+        # Container still running. Must be a timeout.
+        msg = "Timed out waiting for container READY"
+        stdout = self.stdout
+        if stdout:
+            # Juuuuuuust in case it appeared since we last checked
+            if 'READY' in stdout:
+                return
```

Comment:
Okay, I think I follow.  This will also help if the timeout used is so short, that ``stdout`` isn't actually ever checked.  

> preserve the last-checked stdout 

Having actually written something that solves that, I know it's non-trivial.  Still, un-helpful error messages from test code, shouldn't be solved in generic code (as annoying as they are).  That just moves the problem and introduces a case where it's possible to exceede the timeout, but still pass the test  I agree though that tracking examined lines isn't easy, and duplicating may be the best option.

What about making docker do the 'already seen' legwork by tracking the timestamps (since logging a line requires time):

```python
first_timestamp = last_timestamp = now()  # there's a function for this somewhere
n_lines=0
...
while time.time() <= end_time:
    ...
        logs = DockerCmd(self.subtest, 'logs', [, "--timestamps", "--since", last_timestamp, "--tail", "1", self.container_id])
        if logs:
            last_timestamp = # parse it out
            n_lines += 1
```

Then to produce a helpful error message, run ``docker logs --since=%{first_timestamp} --tail=%{n_lines}``.

or something to that effect.  Could that work?

I'm okay with the duplication, my main fear is creating of a API method with a kind of fuzzy-timeout.  A seriously loaded-down system can produce a lot of fuzz :grimacing: 

OTOH, if it's going to block you for more than a few hours, forget it, we can just merge this as-is.